### ISSUE TITLE
Incompatibility with kombu 4.0

### ISSUE THREAD

**Issue by adwhit** [nameko/nameko#378]
When running the pub/sub [example](http://nameko.readthedocs.io/en/stable/built_in_extensions.html#events-pub-sub) from the docs, I get the following error:

`AttributeError: 'Connection' object has no attribute 'exchange_declare'`

Full traceback is here:

https://gist.github.com/adwhit/5007a304d68891ead581e4c9035db205

Nameko version 2.4.2, Kombu version 4.0.0

If I drop down to Kombu 3.0.37, it works. Therefore I assume the problem is either with Kombu or the Nameko-Kombu interface.

*Edit* - I see you have recently updated setup.py to use a version < 4. Feel free to close this issue! I note that every other feature seemed to be working correctly so hopefully there will not be too much work to get 4.0 working.

**Comment by mattbennett** [comment]
Nameko 2.4.3 has now been released, which pins kombu back down to a compatible version.

I'll leave this issue open until we fix the compatibility issue. Thanks for reporting!

**Comment by davidszotten** [comment]
started looking at kombu 4 compatibility. in praticular, it switches to (py-)amqp 2.0 (kombu 3 pins amqp below 2). 2.0 is pretty nice, e.g. has much better error messages for connection issues, and nicer basic return support.

it might be a bit tricky to support both versions, at least without version and/or feature sniffing, which makes coverage tricky (or requires suppressing cov requirements) [should we consider running a single travis job so that we can combine coverage reports e.g. from different python versions]

i guess the argument can be made that (unlike other requirements like 'requests'), nameko "owns" the kombu requirement (i.e. we wouldn't expect user code to depend directly on kombu), so we could decide to (bump our version and) just require kombu 4. not sure how i feel about that, esp. given how recently it came out. i don't really want to have to start maintaining multiple minor versions concurrently (i.e. nameko 3.x with kombu 4 and nameko 2.x with kombu 3)

thoughts and other considerations welcome

**Comment by mattbennett** [comment]
I agree that we should avoid supporting multiple versions of kombu concurrently, and with the argument that nameko "owns" kombu as a dependency. I would be fine upgrading to support kombu 4.0 only after it's shown itself to be reasonably stable.

However, I don't think we really need kombu at all. We only use the AMQP component, and we've had [discussions](https://groups.google.com/forum/#!topic/nameko-dev/WZ-xRYd2-vw) in the past about how kombu restricts us in some ways.

I wonder how much work it'd be to change the AMQP components to use py-amqp (2.0) or another library directly instead of kombu, at least compared to the work required to upgrade to 4.x.

**Comment by mattbennett** [comment]
Also I guess there's no great urgency here. If nameko "owns" the kombu dependency it's fine for us to keep it pinned down to 3.x for the foreseeable future.

**Comment by davidszotten** [comment]
agree that we are in no rush, though it's nice to be compatible with the latest versions of libraries, to be sure to get security updates and bug fixes (not sure what kombu's policy is, e.g. if 3.x will stay maintained at all)

**Comment by levchik** [comment]
I don't know if my problem deserves separate issue, but I have a problem now, which seems to be related to this topic. I have latest celery (4.1.0) in project which depends on kombu 4.1.0. Since I have also nameko, I can't use pytest (and probably code that invokes calls to the nameko microservice, haven't tested yet), it just fails with:

```python
Traceback (most recent call last):
  File "/usr/lib/python3.6/site-packages/pluggy/__init__.py", line 397, in load_setuptools_entrypoints
    plugin = ep.load()
  File "/usr/lib/python3.6/site-packages/pkg_resources/__init__.py", line 2404, in load
    self.require(*args, **kwargs)
  File "/usr/lib/python3.6/site-packages/pkg_resources/__init__.py", line 2427, in require
    items = working_set.resolve(reqs, env, installer, extras=self.extras)
  File "/usr/lib/python3.6/site-packages/pkg_resources/__init__.py", line 875, in resolve
    raise VersionConflict(dist, req).with_context(dependent_req)
pkg_resources.VersionConflict: (kombu 4.1.0 (/usr/lib/python3.6/site-packages), Requirement.parse('kombu<4,>=3.0.1'))

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/bin/pytest", line 11, in <module>
    sys.exit(main())
  File "/usr/lib/python3.6/site-packages/_pytest/config.py", line 50, in main
    config = _prepareconfig(args, plugins)
  File "/usr/lib/python3.6/site-packages/_pytest/config.py", line 160, in _prepareconfig
    pluginmanager=pluginmanager, args=args)
  File "/usr/lib/python3.6/site-packages/pluggy/__init__.py", line 617, in __call__
    return self._hookexec(self, self._nonwrappers + self._wrappers, kwargs)
  File "/usr/lib/python3.6/site-packages/pluggy/__init__.py", line 222, in _hookexec
    return self._inner_hookexec(hook, methods, kwargs)
  File "/usr/lib/python3.6/site-packages/pluggy/__init__.py", line 216, in <lambda>
    firstresult=hook.spec_opts.get('firstresult'),
  File "/usr/lib/python3.6/site-packages/pluggy/callers.py", line 196, in _multicall
    gen.send(outcome)
  File "/usr/lib/python3.6/site-packages/_pytest/helpconfig.py", line 68, in pytest_cmdline_parse
    config = outcome.get_result()
  File "/usr/lib/python3.6/site-packages/pluggy/callers.py", line 76, in get_result
    raise ex[1].with_traceback(ex[2])
  File "/usr/lib/python3.6/site-packages/pluggy/callers.py", line 180, in _multicall
    res = hook_impl.function(*args)
  File "/usr/lib/python3.6/site-packages/_pytest/config.py", line 943, in pytest_cmdline_parse
    self.parse(args)
  File "/usr/lib/python3.6/site-packages/_pytest/config.py", line 1108, in parse
    self._preparse(args, addopts=addopts)
  File "/usr/lib/python3.6/site-packages/_pytest/config.py", line 1071, in _preparse
    self.pluginmanager.load_setuptools_entrypoints('pytest11')
  File "/usr/lib/python3.6/site-packages/pluggy/__init__.py", line 402, in load_setuptools_entrypoints
    "Plugin %r could not be loaded: %s!" % (ep.name, e))
pluggy.PluginValidationError: Plugin 'pytest_nameko' could not be loaded: (kombu 4.1.0 (/usr/lib/python3.6/site-packages), Requirement.parse('kombu<4,>=3.0.1'))!
```

When lowering down kombu to 3.0.17 i get error from celery (something about imports and other errors as well).

Am I missing something obvious or I just can't use nameko and celery in one project due to kombu versions mismatch (can you suggest something in this case)?

**Comment by levchik** [comment]
To the continuation of my previous post, actual code running nameko instruction doing just fine. Successfully sent an event to my microservice from django. Pytest's still an issue though.

**Comment by bweaver-cxp** [comment]
Can we get an update on this?

We're trying to use amqps, which is not supported in kombu <4.0.0, which we've sort of gotten around by using amqp with `?ssl=true` at the end. It mostly works, but it spits out a bunch of errors and it's causing issues in our automated deployments (connections aren't being reliably closed when our containers are replaced, causing our RabbitMQ to slowly run out of memory). We've forked nameko and upgraded the dependency to use kombu >=4.0.0 and switched our uris to use amqps and everything seems to be working great (but of course the nameko tests don't pass).

We need SSL at every level due to HIPAA requirements. We might just run a parallel nameko repository and periodically merge official changes into it, with the only difference being the kombu version. I don't love that solution, but this is a pretty significant issue for us.

**Comment by mattbennett** [comment]
I would still like to ditch kombu in favour of a pure AMQP library, and am getting closer to that point in a couple of branches on my fork.

In the mean time though, I think we should look at upgrading to kombu 4 and dropping support for older versions. If anyone here has the time and inclination to experiment with that I'd be receptive to a pull request.

@bweaver-cxp does the change in #524 help you?

**Comment by bweaver-cxp** [comment]
It doesn't seem to, but I might be using it wrong. What should I have under AMQP_SSL in my config?

The error we fixed by upgrading to kombu 4.0.0 is this:
```Exception ignored in: <bound method _AbstractTransport.__del__ of <amqp.transport.SSLTransport object at 0x7f91ecd06940>>
Traceback (most recent call last):
  File "/usr/local/lib/python3.6/site-packages/amqp-1.4.9-py3.6.egg/amqp/transport.py", line 116, in __del__
    self.close()
  File "/usr/local/lib/python3.6/site-packages/amqp-1.4.9-py3.6.egg/amqp/transport.py", line 141, in close
    self._shutdown_transport()
  File "/usr/local/lib/python3.6/site-packages/amqp-1.4.9-py3.6.egg/amqp/transport.py", line 217, in _shutdown_transport
    self.sock = unwrap()
  File "/usr/local/lib/python3.6/site-packages/eventlet-0.21.0-py3.6.egg/eventlet/green/ssl.py", line 265, in unwrap
    super(GreenSSLSocket, self).unwrap))
  File "/usr/local/lib/python3.6/site-packages/eventlet-0.21.0-py3.6.egg/eventlet/green/ssl.py", line 109, in _call_trampolining
    return func(*a, **kw)
  File "/usr/local/lib/python3.6/ssl.py", line 1055, in unwrap
    raise ValueError("No SSL wrapper around " + str(self))
ValueError: No SSL wrapper around <eventlet.green.ssl.GreenSSLSocket fd=6, family=AddressFamily.AF_INET, type=2049, proto=6, laddr=('172.21.0.8', 50800)>
```

**Comment by mattbennett** [comment]
Oh interesting. I see this same error in the branch in #524, but hadn't noticed the negative side-effects. I guess there's a bug in Eventlet's green SSL implementation, but it may also be in pyamqp. Kombu 4 jumps to pyamqp 2.0, either fixing or avoiding this problem.

If we can identify and work around this bug perhaps the 4.0 upgrade isn't required.

#524 assumes you want to do certificate verification of the client, which I will relax. For now, the `AMQP_SSL` params need to be:

```
AMQP_SSL:
    ca_certs: <certificate-authority-cert-location>
    certfile: <client-cert-location>
    keyfile: <client-key-location>
```

**Comment by mattbennett** [comment]
Actually the branch at #524 doesn't require you to provide client certificates. You can specify `AMQP_SSL ` as just:

```
AMQP_SSL: True
```

The "No SSL wrapper" error can be reproduced by running the following a few times, even without eventlet:

```
from kombu import Connection

conn = Connection("amqp://guest:guest@localhost:5671/", ssl=True)
conn.connect()
assert conn.connected
conn.release()
```

So the bug is plainly in pyamqp. I looked for an issue there but haven't found one.

**Comment by bweaver-cxp** [comment]
We don't know for sure that that particular error is causing our unwanted behavior, we just know that when we don't use SSL (or when we used the most recent Kombu) we don't see our issue, and that correlates with not getting that error. And our problem is only in a very specific environment; it doesn't happen locally or in manual starting and stopping of (Docker) containers on AWS, just in automated deployments on AWS. It's hard to pin down exactly what/where/why the problem is.

I tried both of those configuration options, not realizing that this branch wasn't meant to fix the "No SSL wrapper" error.

**Comment by mattbennett** [comment]
This traceback is fixed in pyamqp 2.0.3 by this commit: https://github.com/celery/py-amqp/commit/7db1b2d4a5cfc546ab941697ae7035ff34645a5a#diff-e3b09014ce12269dff5eece1f9f7c082

I think that `__del__` is sometimes called by the garbage collector after the connection has already been unwrapped. If I comment out the `conn.release()` I can no longer reproduce the error.

Unfortunately I think this means it's unlikely to be the cause of whatever is leaving your connections open. I have just backported the fix into #524 so you can try and see?

**Comment by bweaver-cxp** [comment]
#524 does fix the SSL wrapper error and our connections are still not closing. So I have no idea what the issue or fix is, except that SOME commit between py-amqp v1.4.9 and v2.2.2 must do the trick. I might try a deployment at each version to narrow down the search.

It's nice to have all those logs cleaned up though. Makes it easier to see what's happening.

**Comment by asyncee** [comment]
Can someone clarify is there are still major incompatibility with kombu>4.0? We are using celery, which depends on kombu 4.x. This issue hardly influence the architectural choices, because it can not be used concurrently with celery.

Thanks.

**Comment by asyncee** [comment]
We are currently in a situation where monolithic application is using celery and in a process of migration to soa/microservices architecture. We had two requirements:

1. use flask-nameko or
2. share a codebase in one environment 

We can not split environments due to current architecture. Anyway, this approach introduces unnecessary complexity to support and deploy.

I think it would be great to freeze development of 2.x version and start working on new 3.x that will be backwards incompatible (possibly).

**Comment by mattbennett** [comment]
Fixed in #564

**Comment by davidszotten** [comment]
for future reference i wonder if we want to hold off closing issues until we _release_ a version with the fix
